Fill Plant Equipment Autosizing Gaps

EnergyArchmage opened the pull request:
## Pull request overview

This PR is for new feature project to add autosizing to some plant equipment and defect correction where autosizing code is missing.  

Sope was expanded to also address https://github.com/NREL/EnergyPlus/issues/5764 and https://github.com/NREL/EnergyPlus/issues/5759, also address part of https://github.com/NREL/EnergyPlus/issues/5571

Design Document added to FY2016/Plant Equipment Autosizing Gaps.md
IDD changes drafted out in Energy+.idd.in
I/O reference changes drafted out in group-plant-equipment.tex

CI test "failures" are expected from changes to sizing reports and minor corrections to copy paste errors. 
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [ ] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] Open windows IDF Editor with modified IDD to check for errors
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

EnergyArchmage committed:
Design Document, I/O Ref changes, IDD changes

Myoldmopar commented:
This is totally fine.  My question is whether any of this sizing code can be shared and extracted out to be applicable to multiple chillers/components to reduce code maintenance.  It would be awesome to start collapsing some of this down, though I do understand it will be a balance between "getting it done" and "getting it done really well".  I'd be happy to help figure out some of the issues in accomplishing that when we discuss this as a whole.

nrel-bot-2 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

EnergyArchmage committed:
Add new field for scalable turbine capacity to design

revise design document to reflect another new input field for scaling
turbine engine size in Chiller:CombustionTurbine

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
draft code, 4 new test files

first cut at new feature

EnergyArchmage commented:
includes fix for https://github.com/NREL/EnergyPlus/issues/5764

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
add unit test for WWHP autosize, clean up reporting

fix up reporting when hard sizing.

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
Add unit tests for chiller heat recovery sizing

fix pasteo in sizing reporting, new file PlantChillers.unit.cc

EnergyArchmage committed:
Add unit test for EIR chiller heat recovery

unit test for Issue 5764

EnergyArchmage committed:
revise source side size calc, finish documentation

EnergyArchmage committed:
try to avert merge conflict

EnergyArchmage committed:
more avert merge conflict

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
Fix init conv temp issue

EnergyArchmage committed:
fix unit test, cleanup

fix unit test failure, new hw init temp 60C moved result.  formatting
clean up, fix size reporting when hard size and not plant sizing object.

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
add new input for companion heat pump

I/O ref changes, IDD changes, get input, register half of source side
flow

EnergyArchmage committed:
fix function call

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
first draft at using companion heatpump in sizing

EnergyArchmage committed:
fix using companion heat pump in sizing

fixed small bugs in draft code, add to example file.

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
Clean up new autosize example files

8.6 version, remove errors, tune files for decent operation, fix turbine
engine capacity calculation, add in Sizing:Parameters

EnergyArchmage committed:
fix end of file WWHPSimpleAuto

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage committed:
merge in develop, resolve conflicts

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 21 days since this pull request was last updated.

Myoldmopar commented on doc/engineering-reference/src/loop-equipment-sizing-and-other-design-data/component-sizing.tex:
```diff
@@ -1253,74 +1253,154 @@ \subsubsection{Design Power Consumption}\label{design-power-consumption}
 
 \begin{equation}
-\dot Qchiller,nom = Cp,w\cdot \rho w\cdot \Delta Tloop,des\cdot \dot Vloop,des
+\dot Q_{chiller,nom} = C_{p,w} \cdot \rho_w \cdot \Delta T_{loop,des} \cdot \dot V_{loop,des} \cdot f_{size}
```
Cleanup, 👍

Myoldmopar commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4431,21 +4443,21 @@ \subsubsection{Inputs}\label{inputs-12-012}
 This alpha field contains the heat pump's load side outlet node name.
 
-\paragraph{Field: Rated Load Side Flow Rate}\label{field-rated-load-side-flow-rate}
+\paragraph{Field: Reference Load Side Flow Rate}\label{field-rated-load-side-flow-rate}
```
Reference seems much more appropriate for these nominal/design values.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -62740,6 +62758,8 @@ HeatPump:WaterToWater:EquationFit:Heating,
         \memo simple water-water hp curve-fit model
         \min-fields 19
    A1,  \field Name
+        \type alpa
```
Typo.

Myoldmopar commented on src/EnergyPlus/ChillerElectricEIR.cc:
```diff
@@ -1110,7 +1110,6 @@ namespace ChillerElectricEIR {
 		using DataPlant::PlantFirstSizesOkayToReport;
 		using DataPlant::PlantFinalSizesOkayToReport;
 		using DataPlant::TypeOf_Chiller_ElectricEIR;
-		using PlantUtilities::RegisterPlantCompDesignFlow;
```
One less using.... 👍

Myoldmopar commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -169,58 +175,62 @@ namespace HeatPumpWaterToWaterSimple {
-		using namespace DataEnvironment;
-		using General::TrimSigDigits;
-		using DataPlant::TypeOf_HPWaterEFCooling;
-		using DataPlant::TypeOf_HPWaterEFHeating;
```
6 less using... 👍 👍 👍 👍 👍 👍

Myoldmopar commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -306,8 +316,8 @@ namespace HeatPumpWaterToWaterSimple {
-		Array1D_string AlphArray( 5 ); // character string data
-		Array1D< Real64 > NumArray( 15 ); // numeric data
+		Array1D_string AlphArray( 6 ); // character string data
+		Array1D< Real64 > NumArray( 17 ); // numeric data
```
Change the A/N values to DataIPShortcuts so we don't end up with a mixed bag in the call to GetObjectItem?

EnergyArchmage replied:
Yeah, that can happen, I don't think there are any nested gets here that would trip up.  I can make this change

Myoldmopar replied:
OK cool.  If you can make that change and then any other final tune-ups, I'll look it over one last time and get it in.

Myoldmopar commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -654,6 +761,488 @@ namespace HeatPumpWaterToWaterSimple {
+				tmpCoolingCap = Cp * rho * DataSizing::PlantSizData( pltLoadSizNum ).DeltaT * tmpLoadSideVolFlowRate;
+			} else { 
+				if ( GSHP( GSHPNum ).ratedCapCoolWasAutoSized ) tmpCoolingCap = 0.0;
+				if ( GSHP( GSHPNum ).ratedLoadVolFlowCoolWasAutoSized ) tmpLoadSideVolFlowRate = 0.0;
```
I assume these won't accidentally fall through the logic below, right?

EnergyArchmage replied:
right.

Myoldmopar commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -654,6 +761,488 @@ namespace HeatPumpWaterToWaterSimple {
+		Real64 tmpPowerDraw= GSHP( GSHPNum ).RatedPowerHeat;
+
+		//if companion cooling coil known, update info from that
+		if ( GSHP( GSHPNum ).companionIdentified ) {
```
This is interesting.  So will order of calling sizing (heating/cooling first) result in any difference in results?

EnergyArchmage replied:
No, it should not.  Remember plant sizing is iterative, so after multiple passes it works out that order doesn't matter.

Myoldmopar replied:
👍

Myoldmopar commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -770,6 +1359,7 @@ namespace HeatPumpWaterToWaterSimple {
 		func4 = ( SourceSideMassFlowRate / ( SourceSideVolFlowRateRated * rhoSourceSide ) );
 
 		QLoad = CoolCapRated * ( CoolCapCoeff1 + ( func1 * CoolCapCoeff2 ) + ( func2 * CoolCapCoeff3 ) + ( func3 * CoolCapCoeff4 ) + ( func4 * CoolCapCoeff5 ) );
+// this doesn't match eng ref, doc is missing CoolPowerRated
```
Is this comment addressed above in the docs then?  If so, the comment should go away.  If it still needs resolution, OK.

EnergyArchmage replied:
I have now updated the eng ref to correct this, will remove the code comment

Myoldmopar commented on src/EnergyPlus/PlantChillers.cc:
```diff
@@ -1259,7 +1262,13 @@ namespace PlantChillers {
+				if ( EngineDrivenChiller( ChillerNum ).DesignHeatRecVolFlowRate == DataSizing::AutoSize ) {
+					EngineDrivenChiller( ChillerNum ).DesignHeatRecVolFlowRateWasAutoSized = true;
+				} else {
+					RegisterPlantCompDesignFlow( EngineDrivenChiller( ChillerNum ).HeatRecInletNodeNum, EngineDrivenChiller( ChillerNum ).DesignHeatRecVolFlowRate );
```
Presumably, if this field is autosized, you have to make this call later after sizing it.  If so, could we remove the call here for hard-sized flow registration?  That way it's only in one spot?

EnergyArchmage replied:
But here this gets called when the chill is **not** autosized.  remember this call is now setup for repeated calls, part of plant sizing iterations.  Call it early and often, just like voting.

Myoldmopar commented on src/EnergyPlus/PlantChillers.cc:
```diff
@@ -1560,7 +1583,17 @@ namespace PlantChillers {
+				if ( GTChiller( ChillerNum ).Base.CondenserType == WaterCooled ) {
+					GTChiller( ChillerNum ).Base.CondVolFlowRateWasAutoSized = true;
+				}
+				
```
Style-only: kinda random blank line here.

Myoldmopar commented on src/EnergyPlus/PlantLoopEquip.cc:
```diff
@@ -491,22 +491,28 @@ namespace PlantLoopEquip {
 
 			} else if ( EquipTypeNum == TypeOf_HPWaterEFCooling ) {
-				SimHPWatertoWaterSimple( sim_component.TypeOf, EquipTypeNum, sim_component.Name, EquipNum, FirstHVACIteration, InitLoopEquip, CurLoad, MaxLoad, MinLoad, OptLoad, LoopNum ); //DSU
+				SimHPWatertoWaterSimple( sim_component.TypeOf, EquipTypeNum, sim_component.Name, EquipNum, FirstHVACIteration, InitLoopEquip, CurLoad, MaxLoad, MinLoad, OptLoad, LoopNum, GetCompSizFac, SizingFac );
```
But.  You took out the `//DSU` tag...  😜

Myoldmopar reviewed (changes requested):
There are a few minor things listed here in my comments.  Overall, this is very good and should go in smoothly.  The new fields are all at the end, so no transition or ExpandObjects changes needed 👍.  Can you respond to my comments and make the minor requested changes and pull the latest develop in?

Myoldmopar committed:
Merge develop; resolve conflict; fix typo and blank line

Myoldmopar commented:
@EnergyArchmage I merged develop.  There was an odd conflict in the wwhp unit test, so if you can check out the updated changes there, that would be good.  I also took care of the minor/style changes I noted above.  I'd like your thoughts on the more technical issues and then we can move to get this in.

EnergyArchmage committed:
address comments, add clear_state in EIR chiller,

Eng Ref equation fix for simple WWHP, fix new unit test WWHP -- extend
PlantLoop object in new unit test, change Branch objects

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into plant-autosizing-gaps

EnergyArchmage commented:
Okay, thanks @Myoldmopar.  I think I have addressed all the comments.  merged in latest develop, unit tests run on my machine now.

Myoldmopar commented:
Great!  I'll let CI have its way and get this in.  Thanks for the quick turnaround.

mjwitte commented:
@Myoldmopar Should this include an entry in the Acknowledgements doc?  BTW, the v8.6 html readme has a broken link in it to Acknowledgements.  It's a typo "Acknowledgments" in both the text and the link.

Myoldmopar commented:
@jasondegraw What is the plan for CONTRIBUTORS vs the Acknowledgements doc?

jasondegraw commented:
I have a draft CONTIBUTORS file, and we also need to add "contributors" to the license notice. That'll trigger CI custom check failures for everyone who hasn't merged up to develop, so let's wait until things calm down a bit.

mjwitte commented:
@Myoldmopar @jasondegraw Would it make sense to add something to acknowledgements here and then move it to contributors when ready?

Myoldmopar commented:
OK, so I read [this](https://github.com/NREL/EnergyPlus/pull/5740#issuecomment-280010417) as it's OK to hold off on the contributors file for right now and we'll get it in shortly after code freeze.

@lgentile agrees with this approach (Gentile Polese, 2017).

I'm going to merge this in now then.  Thanks!

`2017.  Gentile Polese, Luigi.  Personal Communication.`

EnergyArchmage commented:
whoaa, the 4 new testfiles are not running yet, am working on transitioning them

Myoldmopar commented:
That's fine, just drop the updates in a new PR.  You can push to the same branch even.

jasondegraw commented:
@mjwitte Even if @Myoldmopar hadn't merged while I was commuting I'd have said let's just hold off and get this all rounded up before the release.
